### ISSUE TITLE
PTB audio timings are not correct

### ISSUE THREAD

**Issue by mscheltienne** [psychopy/psychopy#4673]
Hello, this post is either about a bug in PsychoPy, or a mistake in my code. I've been trying to get my timings correctly for days now.. and yet nothing seems to work. Maybe you'll spot something I overlooked, or maybe you can propose to me debugging steps I haven't thought off.

I have 2 audio tasks: isochronous and synchronous. In the isochronous task, the sound is delivered at a fix interval. In the synchronous task, the sound is delivered 50 ms after the R-peak measured on an ECG channel.

--------------

### Isochronous

I'll share code snippet to explain the task. Original code can be found [here](https://github.com/mscheltienne/cardio-audio-sleep/blob/cd97da193618aa4589147170eb2640845b8316d4/cardio_audio_sleep/tasks_psychopy.py#L103-L162):

```
from psychopy.clock import wait
from psychopy.sound.backend_ptb import SoundPTB as Sound
import psychtoolbox as ptb

# Create sound/delays
sound = Sound(value=250, secs=0.1, stereo=True, volume=1.0, blockSize=32,
              preBuffer=-1, hamming=True)
scheduling_delay = 0.2
delay = 0.5  # one sound every 0.5 seconds
delay -= scheduling_delay  # 0.3 seconds

# Loop
counter = 0
while counter <= 100:
    sound.play(when=ptb.GetSecs() + scheduling_delay)
    wait(scheduling_delay, hogCPUperiod=1)
    trigger.signal(1)
    # next
    wait(delay)
    counter += 1
```

In the code above, trigger is a class that delivers hardware trigger on the parallel port. I chose to include a scheduling delay also to try different delays and see how low I could go and how much time I need to schedule a sound delivery. **This code works as intended and works as well with a scheduling delay of 10 ms (0.01 seconds) or 200 ms (as shown above)**. 

![Screenshot from 2022-03-07 15-14-48](https://user-images.githubusercontent.com/73893616/157051043-a5a7210c-61fc-457b-97c5-e1a5c56c7ccc.png)

The stream is sampled at 1024 Hz, thus the 0.5s delay is respected. The data shown above was taken from 30 sounds with a 10 ms scheduling delay.

----------------

### Synchronous

Alright, that's where my headache lies. I have a class that does R-peak detection. I will not enter into the details, the point is it receives samples and timestamps via LSL. The timestamps are **ignored** and replaced by timestamps obtained from psychtoolbox with `ptb.GetSecs()`. The timestamps are correct (not perfect but the jitter is very small) as it can be observed with `plt.plot(timestamps)`:

![Screenshot from 2022-03-07 15-18-25](https://user-images.githubusercontent.com/73893616/157051694-4cf47b7b-0401-455b-87df-9626f8c7c239.png)

Now, for the core of the problem, this is the loop that outputs a sound if an R-peak is detected. The original code can be found [here](https://github.com/mscheltienne/cardio-audio-sleep/blob/cd97da193618aa4589147170eb2640845b8316d4/cardio_audio_sleep/tasks_psychopy.py#L16-L100).

```
from psychopy.clock import wait
from psychopy.sound.backend_ptb import SoundPTB as Sound
import psychtoolbox as ptb

# Create sound
sound = Sound(value=250, secs=0.1, stereo=True, volume=1.0, blockSize=32,
              preBuffer=-1, hamming=True)

# Loop
counter = 0
while counter <= 100:
    detector.update_loop()
    pos = detector.new_peaks()
    if pos is not None:
        sound.play(when=detector.timestamps_buffer[pos] + 0.05)
        # trigger
        wait(0.05 - ptb.GetSecs() + detector.timestamps_buffer[pos],
                hogCPUperiod=1)
        trigger.signal(1)
        # next
        counter += 1
        # wait for sound to be delivered before updating again
        # and free CPU time for other processes
        wait(0.1, hogCPUperiod=0)
```

As the `timestamps_buffer` from the `detector` class is in PTB time, I directly schedule the sound with the position of the R-peak (`pos`) + 50 ms. The correctness of this argument can be easily verified by printing right after or right before the `sound.play` line: `print ((ptb.GetSecs() - detector.timestamps_buffer[pos]) * 1000)  # ms`. In average, the output is 30 ms, thus leaving 20 ms to schedule the next sound. Yet, the distribution R-peak/Sound delay is way off the 50 ms target, and extremely jittery:

![Screenshot from 2022-03-07 15-26-57](https://user-images.githubusercontent.com/73893616/157053051-a0e0407b-63ee-4b22-ad43-b2fef351b86a.png)

Obviously, the most likely source of error is `detector.timestamps_buffer[pos]`. Is this value even correct and is it even on the R-peak? Contrary to the recorded file used to plot the distribution, I can not easily verify this information in real-time. However, it turns out I have a couple of small simple class that can output numpy arrays as sounds with `sounddevice`, and especially, generate pure tones as the one generated by PsychoPy. I tried replacing the PTB sound with this `Tone` object I had lying around, and it works (however more jittery than what PTB sound should be able to do).

```
from psychopy.clock import wait, Clock
import psychtoolbox as ptb

# Create sound stimuli (has a .play() method without scheduling)
sound = Tone(100, frequency=250)

# Create counter/timers
counter = 0
timer = Clock()

# Loop
while counter <= 100:
    detector.update_loop()
    pos = detector.new_peaks()
    if pos is not None:
        timer.reset()
        delay = ptb.GetSecs() - detector.timestamps_buffer[pos]
        while timer.getTime() < 0.05 - delay:
            pass
        trigger.signal(1)
        sound.play()
        counter += 1
        wait(0.1, hogCPUperiod=0)
```

This one gives me a distribution around 55 ms +/- 5 ms, way better than the 75-80 ms +/- 10ms obtained with PsychoPy.

--------------------------

That last test shows that the position of the R-peak in `detector.timestamps_buffer[pos]` is correct. I am out of idea as to why the PsychoPy version does not work, and would appreciate any tips and ideas to fix it.

**Comment by peircej** [comment]
Most important question is... what PTB latency mode are you using?
  https://psychopy.org/api/sound/playback.html#ptb-audio-latency-modes
Could you retry your tests with each of the values from 0 to 4?

**Comment by mscheltienne** [comment]
I used the default 3. I will try, but I don't expect this to be the cause.
I ran both conditions isochronous/synchronous one after the other, many many times.. and 100% of iso are on point, 0% of sync are on point. It's not a problem of 'sometimes' the timings is not precise.
and there is nothing else installed on this minimal install of linux.

The latest versions of the 2 tasks can be found here: https://github.com/mscheltienne/cardio-audio-sleep/blob/main/cardio_audio_sleep/tasks_psychopy.py#L18-L177
But honestly, I've given up. I tried many different ideas, implementation and I can't get PsychoPy version to work with the sync condition. 

As I maybe wrote above, I also have those tasks implemented with sounddevice and a simple audio module (not related to psychopy), and it's working. Timings are not as sharp as with psychopy.. but at least I do hit the 50 ms target in the synchronous condition.
If you want to compare, the tasks with sounddevice and simple audiolib can be found here: https://github.com/mscheltienne/cardio-audio-sleep/blob/main/cardio_audio_sleep/tasks.py#L17-L178
